### DISCUSSION TITLE
How can i read and mofify parameter when REGISTER？

### DISCUSSION THREAD

**Discussion by ChuLiqiang** [resiprocate/resiprocate#287] [Q&A] [answered]
I want add GB/35114 to resiprocate. UAC and UAS have some interface to modify and read new parameters?
35114 let me to:
(1)add authorization parameter() when send first REGISTER message(UAC）
Authorization:Capability algorithm=
"A:SM2;H:SM3;S:SM1/OFB/PKCS5,SM1/CBC/PKCS5,SM4/OFB/PKCS5,SM4/CBC/PKCS5;SI:SM3-SM2",
keyversion="" 
(2)modify WWW-Authenticate parameter in 401 message (UAS)
WWW-Authenticate:Unidirection algorithm="A:SM2;H:SM3;S:SM1/OFB/PKCS5;SI:SM3-SM2",random1="YTHFjdo37RiZvck3nNoqTA=="
(3)modify authorization parameter when send second REGISTER message(UAC)
Authorization:Unidirection random1="",random2="",serverid="服务器编码",sign1="",algorithm="A:SM2;H:SM3;S:SM1/OFB/PKCS5;SI:SM3-SM2"
(4) add SecurityInf parameter when send 200 ok message (UAS)
SecurityInf:Unidirection cryptkey="",algorithm="A:SM2;H:SM3"

**Comment by sgodin** [comment]
I don't know the specifics of GB/35114, however this may help....

Here are some resources to check out:
https://github.com/resiprocate/resiprocate/wiki/Use-Overview
https://github.com/resiprocate/resiprocate/wiki/Use-Overview#Parameters

Also DUM has a framework for implementing authentication.  See DialogUsageManager::setClientAuthManager and setServerAuthManager. 

Best Regards,
Scott

**Reply by ChuLiqiang** [reply]
WWW-Authenticate modify parameter type and name for example:
WWW-Authenticate:Bidirection algorithm="A:SM2;H:SM3;S:SM1/OFB/PKCS5;SI:SM3-SM2",random1="YTHFjdo37RiZvck3nNoqTA=="
in resip WWW-Authenticate define by:
WWW-Authenticate | h_WWWAuthenticates | Auths

How can i modify this by elegant way.I wan not modify WWW-Authenticate to StringCategory in resip source code.

**Comment by sgodin** [comment, marked as answer]
So Auth scheme is "Bidirection" instead of Digest, the algorithm parameter
is just a string, so it can contain your content.  The only oddity is the
random1 parameter, which you could use an extension parameter for without
needing to modify the resip stack.

See the info in the Extension parameter section of this page:
https://github.com/resiprocate/resiprocate/wiki/Use-Overview#Extension_parameters

Scott

**Reply by ChuLiqiang** [reply]
why we use Decorator in send second reg message. can we add Auth value in addAuthentication function? it olny a Design Patterns?not have other reason?
![image](https://user-images.githubusercontent.com/2747722/207292723-4c111646-b365-4258-8df0-25a25913108d.png)

**Comment by sgodin** [comment]
The explanation is in the comment above addOutboundDecorator call.

Auth-int digest authentication requires that a hash of the message body be
used.  An application may be using an outbound decorator to change an SDP
body (ie: update the connection address address to match the interface
selected to send the Sip Message on).  Auth-int would fail if the body hash
was calculated in the addAuthentication method, then later changed.

Scott

**Reply by ChuLiqiang** [reply]
thanks for your help.I will try modify by your comment.if i have result i will share in this discussions.
